diff --git a/prov/sockets/src/sock_av.c b/prov/sockets/src/sock_av.c
index e33e3b883..05d3972a1 100644
--- a/prov/sockets/src/sock_av.c
+++ b/prov/sockets/src/sock_av.c
@@ -441,10 +441,12 @@ static int sock_av_remove(struct fid_av *av, fi_addr_t *fi_addr, size_t count,
 	}
 	fastlock_release(&_av->list_lock);
 
+	fastlock_acquire(&_av->table_lock);
 	for (i = 0; i < count; i++) {
 		av_addr = &_av->table[fi_addr[i]];
 		av_addr->valid = 0;
 	}
+	fastlock_release(&_av->table_lock);
 
 	return 0;
 }
diff --git a/prov/sockets/src/sock_conn.c b/prov/sockets/src/sock_conn.c
index 8c739db74..be080efe2 100644
--- a/prov/sockets/src/sock_conn.c
+++ b/prov/sockets/src/sock_conn.c
@@ -167,8 +167,9 @@ void sock_conn_release_entry(struct sock_conn_map *map, struct sock_conn *conn)
 	ofi_close_socket(conn->sock_fd);
 
 	conn->address_published = 0;
-        conn->connected = 0;
-        conn->sock_fd = -1;
+	conn->av_index = FI_ADDR_NOTAVAIL;
+	conn->connected = 0;
+	conn->sock_fd = -1;
 }
 
 static int sock_conn_get_next_index(struct sock_conn_map *map)
@@ -540,7 +541,11 @@ do_connect:
 
 			ret = poll(&poll_fd, 1, sock_conn_timeout);
 			if (ret < 0) {
-				SOCK_LOG_DBG("poll failed\n");
+				SOCK_LOG_ERROR("poll failed: %s\n",
+					       strerror(ofi_sockerr()));
+				goto retry;
+			} else if (ret == 0) {
+				SOCK_LOG_ERROR("poll timed out\n");
 				goto retry;
 			}
 
@@ -548,21 +553,22 @@ do_connect:
 			ret = getsockopt(conn_fd, SOL_SOCKET, SO_ERROR,
 					 (void*)(&valopt), &lon);
 			if (ret < 0) {
-				SOCK_LOG_DBG("getsockopt failed: %d, %d\n",
-					     ret, conn_fd);
+				SOCK_LOG_ERROR("getsockopt failed: %d, %d\n",
+					       ret, conn_fd);
 				goto retry;
 			}
 
 			if (valopt) {
-				SOCK_LOG_DBG("Error in connection() "
-					     "%d - %s - %d\n",
-					     valopt, strerror(valopt), conn_fd);
+				SOCK_LOG_ERROR("Error in connection() "
+					       "%d - %s - %d\n",
+					       valopt, strerror(valopt),
+					       conn_fd);
 				goto retry;
 			}
 			goto out;
 		} else {
-			SOCK_LOG_DBG("Timeout or error() - %s: %d\n",
-				     strerror(ofi_sockerr()), conn_fd);
+			SOCK_LOG_ERROR("Timeout or error() - %s: %d\n",
+				       strerror(ofi_sockerr()), conn_fd);
 			goto retry;
 		}
 	} else {
@@ -571,6 +577,11 @@ do_connect:
 
 retry:
 	do_retry--;
+	ofi_straddr_log(&sock_prov, FI_LOG_WARN, FI_LOG_EP_CTRL,
+			"Error connecting to ", &addr.sa);
+	SOCK_LOG_ERROR("Connect error%s - %s - %d\n",
+		       do_retry ? ", retrying" : "", strerror(ofi_sockerr()),
+		       conn_fd);
 	if (!do_retry)
 		goto err;
 
@@ -579,8 +590,6 @@ retry:
 		conn_fd = -1;
 	}
 
-	SOCK_LOG_ERROR("Connect error, retrying - %s - %d\n",
-		       strerror(ofi_sockerr()), conn_fd);
         goto do_connect;
 
 out:
diff --git a/prov/sockets/src/sock_ep.c b/prov/sockets/src/sock_ep.c
index 2f1ff9add..18d8dfdd1 100644
--- a/prov/sockets/src/sock_ep.c
+++ b/prov/sockets/src/sock_ep.c
@@ -1786,6 +1786,8 @@ err1:
 
 void sock_ep_remove_conn(struct sock_ep_attr *attr, struct sock_conn *conn)
 {
+	if (attr->cmap.used <= 0 || conn->sock_fd == -1)
+		return;
 	sock_pe_poll_del(attr->domain->pe, conn->sock_fd);
 	sock_conn_release_entry(&attr->cmap, conn);
 }
@@ -1794,7 +1796,7 @@ struct sock_conn *sock_ep_lookup_conn(struct sock_ep_attr *attr, fi_addr_t index
 				      union ofi_sock_ip *addr)
 {
 	int i;
-	uint16_t idx;
+	uint64_t idx;
 	char buf[8];
 	struct sock_conn *conn;
 
@@ -1802,16 +1804,19 @@ struct sock_conn *sock_ep_lookup_conn(struct sock_ep_attr *attr, fi_addr_t index
 
 	conn = ofi_idm_lookup(&attr->av_idm, idx);
 	if (conn && conn != SOCK_CM_CONN_IN_PROGRESS) {
-		if (conn->connected == 0)
-			return NULL;
 		/* Verify that the existing connection is still usable, and
 		 * that the peer didn't restart.
 		 */
-		if (sock_comm_peek(conn, buf, 8) == 0 && conn->connected == 0) {
-			SOCK_LOG_DBG("Disconnected\n");
+		if (conn->connected == 0 ||
+		    (sock_comm_peek(conn, buf, 8) == 0 && conn->connected == 0)) {
+			sock_ep_remove_conn(attr, conn);
+			ofi_straddr_log(&sock_prov, FI_LOG_WARN, FI_LOG_EP_CTRL,
+					"Peer disconnected", &addr->sa);
 			return NULL;
 		}
-		if (conn->av_index == FI_ADDR_NOTAVAIL)
+		if (conn->av_index != FI_ADDR_NOTAVAIL)
+			assert(conn->av_index == idx);
+		else
 			conn->av_index = idx;
 		return conn;
 	}
@@ -1822,18 +1827,21 @@ struct sock_conn *sock_ep_lookup_conn(struct sock_ep_attr *attr, fi_addr_t index
 
 		if (ofi_equals_sockaddr(&attr->cmap.table[i].addr.sa, &addr->sa)) {
 			conn = &attr->cmap.table[i];
-			if (conn->av_index == FI_ADDR_NOTAVAIL)
-				conn->av_index = idx;
 			break;
 		}
 	}
 	if (conn && conn != SOCK_CM_CONN_IN_PROGRESS) {
-		if (conn->connected == 0)
-			return NULL;
-		if (sock_comm_peek(conn, buf, 8) == 0 && conn->connected == 0) {
-			SOCK_LOG_DBG("Disconnected\n");
+		if (conn->connected == 0 ||
+		    (sock_comm_peek(conn, buf, 8) == 0 && conn->connected == 0)) {
+			sock_ep_remove_conn(attr, conn);
+			ofi_straddr_log(&sock_prov, FI_LOG_WARN, FI_LOG_EP_CTRL,
+					"Peer disconnected", &addr->sa);
 			return NULL;
 		}
+		if (conn->av_index != FI_ADDR_NOTAVAIL)
+			assert(conn->av_index == idx);
+		else
+			conn->av_index = idx;
 	}
 	return conn;
 }
@@ -1868,9 +1876,11 @@ int sock_ep_get_conn(struct sock_ep_attr *attr, struct sock_tx_ctx *tx_ctx,
 		ret = sock_ep_connect(attr, av_index, &conn);
 
 	if (!conn) {
-		SOCK_LOG_ERROR("Undable to find connection entry. "
+		SOCK_LOG_ERROR("Unable to find connection entry. "
 			       "Error in connecting: %s\n",
 			       fi_strerror(-ret));
+		ofi_straddr_log(&sock_prov, FI_LOG_WARN, FI_LOG_EP_CTRL,
+				"Unable to connect to", &addr->sa);
 		return -FI_ENOENT;
 	}
 
diff --git a/prov/sockets/src/sock_progress.c b/prov/sockets/src/sock_progress.c
index 151ee503c..e4a0892f1 100644
--- a/prov/sockets/src/sock_progress.c
+++ b/prov/sockets/src/sock_progress.c
@@ -1942,13 +1942,12 @@ static int sock_pe_progress_tx_entry(struct sock_pe *pe,
 		goto out;
 
 	if (sock_comm_is_disconnected(pe_entry)) {
-		SOCK_LOG_DBG("conn disconnected: removing fd from pollset\n");
-		if (pe_entry->ep_attr->cmap.used > 0 &&
-		     pe_entry->conn->sock_fd != -1) {
-			fastlock_acquire(&pe_entry->ep_attr->cmap.lock);
-			sock_ep_remove_conn(pe_entry->ep_attr, pe_entry->conn);
-			fastlock_release(&pe_entry->ep_attr->cmap.lock);
-		}
+		ofi_straddr_log(&sock_prov, FI_LOG_WARN, FI_LOG_EP_DATA,
+				"Peer disconnected: removing fd from pollset",
+				&pe_entry->conn->addr.sa);
+		fastlock_acquire(&pe_entry->ep_attr->cmap.lock);
+		sock_ep_remove_conn(pe_entry->ep_attr, pe_entry->conn);
+		fastlock_release(&pe_entry->ep_attr->cmap.lock);
 
 		sock_pe_report_tx_error(pe_entry, 0, FI_EIO);
 		pe_entry->is_complete = 1;
@@ -2021,13 +2020,12 @@ static int sock_pe_progress_rx_pe_entry(struct sock_pe *pe,
 	int ret;
 
 	if (sock_comm_is_disconnected(pe_entry)) {
-		SOCK_LOG_DBG("conn disconnected: removing fd from pollset\n");
-		if (pe_entry->ep_attr->cmap.used > 0 &&
-		     pe_entry->conn->sock_fd != -1) {
-			fastlock_acquire(&pe_entry->ep_attr->cmap.lock);
-			sock_ep_remove_conn(pe_entry->ep_attr, pe_entry->conn);
-			fastlock_release(&pe_entry->ep_attr->cmap.lock);
-		}
+		ofi_straddr_log(&sock_prov, FI_LOG_WARN, FI_LOG_EP_DATA,
+				"Peer disconnected: removing fd from pollset",
+				&pe_entry->conn->addr.sa);
+		fastlock_acquire(&pe_entry->ep_attr->cmap.lock);
+		sock_ep_remove_conn(pe_entry->ep_attr, pe_entry->conn);
+		fastlock_release(&pe_entry->ep_attr->cmap.lock);
 
 		if (pe_entry->pe.rx.header_read)
 			sock_pe_report_rx_error(pe_entry, 0, FI_EIO);
